torch/distributed/_sharded_tensor/api.py

```diff
@@ -550,9 +559,46 @@ def sharding_spec(self) -> ShardingSpec:
         """
         return self._sharding_spec
 
+    def reshard(self, resharding_spec: ReshardingSpec) -> Union[ShardedTensor, torch.Tensor]:
```

Comment:
@fduwjj @pritamdamania87 I feel like "reshard" seems a bit confusing to user. it seems to me that "output_local_tensors" case we will convert a sharded tensor to a data parallel tensor, is this a "reshard" behavior? or is it a "to_replicated_tensor" or "all_gather_shards"? I feel like we aren't "reshard" the tensor from a user prospective.

Reply:
@wanchaol There are two parts here, the first part is `resharding` which is done as per the provided ShardingSpec. Now the second part is `output_local_tensors`, this essentially means that we want to return the local tensors after this operation. For example, this is very useful in cases where the rest of the code is DataParallel and works only with Tensors and not ShardedTensors. Given that ShardedTensors have a very limited set of ops, resharding and converting to Tensors would be very useful in certain cases.

Reply:
Yeah "resharding" API is necessary I think, what I mentioned is the second part of it, basically `output_local_tensors` is kinda fused with the resharding, but I feel like it's a different behavior and we should separate them tbh. The second part is essentially a conversion from ShardedTensor to Data Parallel tensors, it's not a "resharding" behavior, as the output is not another ShardedTensor, rather, it's a ReplicatedTensor. I can't think of a rational API of this yet, maybe sth like `ShardedTensor.to_replicated()`? I just feel like a fused behavior between these two is a bit confusing.

Reply:
Also if we separate these two behaviors, I think we could get rid of the "ReshardingSpec" and just use ShardingSpec universally, which might be a bit cleaner.

Reply:
For this use case, it is indeed not a resharding strictly speaking. It's not trying to converted to a replicate, neither. But resharding of sharded tensor is still needed as @pritamdamania87 has pointed out. So what if we enable the resharding for Partial Tensor for now and create a `to_dp_tensor` method for sharded tensor? What do you guys think? @wanchaol @pritamdamania87

Reply:
After thinking this for a while, I think this indeed is a cleaner option. We don't need to define a new resharding spec. For the API's naming, I guess this is more like a reverse action of sharding op. (aka. the reverse op of api `sharded_param`.) Maybe let's call it `collect_local_shards` (open for a better name, too) ? We then can use this API for partial tensor. 

So basically, for Partial Tensor or sharded Tensor -> Tensor, we can directly call this API to do the convert. If we want to covert one sharded Tensor to a new sharded tensor, we then need `reshard_output`.

Does this sound clearer and less confusing? @wanchaol and @pritamdamania87

Reply:
Yeah that make sense to me, either `extract_output_tensors` or `collect_local_shards` works for me, agreed we will need to implement build a higher level api on them in the Sharder API. and yeah we don't need ReshardingSpec, could use ShardingSpec universally.

Reply:
`collect_local_shard` sounds good to me. I think it is simpler to have the API deal with just one local shard rather than return `List[Tensor]`, because in most use cases we will actually have just a single Tensor and if we return List[Tensor], there needs to be additional logic somewhere to convert it to a single Tensor for most existing data parallel code.

Reply:
cc @mrshenli If you have any better ideas/suggestions around the API :)